### PR TITLE
Implementing the add-after-delete semantic

### PR DESCRIPTION
**Pull request by Framba-Luca**
PR linked to the discussion #191 

Before merging this PR, the engines that don't implement the **add-after-delete** semantic (i know about tamer and aries) should never return **UNSOLVABLE_PROVEN** but **UNSOLVABLE_INCOMPLETELY**. 

Maybe, a **conservative** check can be done to see if the **add-after-delete** semantic **never impacts the problem**. 
In this case, the engine **might** return **UNSOLVABLE_PROVEN**.

### CODE REVIEW

File: unified_planning/model/action.py [hunk 1]
```diff
@@ -402,16 +418,38 @@ def _add_effect_instance(self, effect: "up.model.effect.Effect"):
...
+                        )
+                        or (
+                            assigned_value.is_real_constant()
+                            and effect.value.is_real_constant()
+                        )
+                        or (
+                            assigned_value.is_object_exp()
+                            and effect.value.is_object_exp()
+                        )
+                    )
```
Comment (alvalentini) [inline unified_planning/model/action.py:440]:
```suggestion
                    and assigned_value.is_constant()
                    and effect.value.is_constant()
```

File: unified_planning/model/action.py [hunk 2]
```diff
@@ -402,16 +418,38 @@ def _add_effect_instance(self, effect: "up.model.effect.Effect"):
...
                 ):
                     raise UPConflictingEffectsException(
                         f"The effect {effect} is in conflict with the effects already in the action."
                     )
-                self._fluents_assigned.add(effect.fluent)
+                if (
+                    assigned_value is None
+                    or effect.value.is_int_constant()
+                    or effect.value.is_real_constant()
+                    or effect.value.is_object_exp()
```
Comment (alvalentini) [inline unified_planning/model/action.py:450]:
```suggestion
                    or effect.value.is_constant()
```

File: unified_planning/model/mixins/timed_conds_effs.py [hunk 3]
```diff
@@ -301,31 +319,56 @@ def _add_effect_instance(
...
+                        )
+                        or (
+                            assigned_value.is_real_constant()
+                            and effect.value.is_real_constant()
+                        )
+                        or (
+                            assigned_value.is_object_exp()
+                            and effect.value.is_object_exp()
+                        )
+                    )
```
Comment (alvalentini) [inline unified_planning/model/mixins/timed_conds_effs.py:343]: Same

File: unified_planning/model/mixins/timed_conds_effs.py [hunk 4]
```diff
@@ -301,31 +319,56 @@ def _add_effect_instance(
...
                 ):
                     raise UPConflictingEffectsException(
                         f"The effect {effect} at timing {timing} is in conflict with the effects already in the action."
                     )
-                fluents_assigned.add(effect.fluent)
+                if (
+                    assigned_value is None
+                    or effect.value.is_int_constant()
+                    or effect.value.is_real_constant()
+                    or effect.value.is_object_exp()
```
Comment (alvalentini) [inline unified_planning/model/mixins/timed_conds_effs.py:353]: Same

**Comment by Framba-Luca** [comment]
@aiplan4eu/engine-developers 
Hi everyone, after this PR will be merged in master the UP will implement the `add-after-delete` semantic.

If your engine does not support it, the `PlanGenerationResult` returned -or anything returned by the engine to the user- might need some modifications; for example if the status is `UNSOLVABLE_PROVEN`, a valid plan might exist when enforcing the `add-after-delete` semantic, therefore it must be changed to `UNSOLVABLE_INCOMPLETELY`.

Should you have any question or issue just let me know by commenting here or via email!

File: unified_planning/model/action.py [hunk 5]
```diff
@@ -402,16 +419,28 @@ def _add_effect_instance(self, effect: "up.model.effect.Effect"):
...
+                        assigned_value.is_constant()
+                        and effect.value.is_constant()
+                        and assigned_value.constant_value()
+                        == effect.value.constant_value()
+                    ):
+                        raise UPConflictingEffectsException(
+                            f"The effect {effect} is in conflict with the effects already in the action."
+                        )
+                else:
+                    self._fluents_assigned[effect.fluent] = effect.value
```
Comment (alvalentini) [inline unified_planning/model/action.py:443]: @Framba-Luca This piece of code is duplicated in 3 different files. Can we refactor it?

**Pull request merged**